tools/ReadConsoleInputStream/Program.cs

```diff
@@ -0,0 +1,151 @@
+            mode &= ~Kernel32.CONSOLE_INPUT_MODE.ENABLE_ECHO_INPUT;
+            mode &= ~Kernel32.CONSOLE_INPUT_MODE.ENABLE_LINE_INPUT;
+
+            if (!Kernel32.SetConsoleMode(handle, mode))
+                throw NativeMethods.GetExceptionForWin32Error(Marshal.GetLastWin32Error());
+
+            // set utf-8 cp
+            if (!Kernel32.SetConsoleCP(65001))
+                throw NativeMethods.GetExceptionForWin32Error(Marshal.GetLastWin32Error());
+
+            if (!Kernel32.SetConsoleOutputCP(65001))
+                throw NativeMethods.GetExceptionForWin32Error(Marshal.GetLastWin32Error());
+
+            // base our provider/consumer on a circular buffer to keep memory usage under control
+            var events = new BlockingCollection<Kernel32.INPUT_RECORD>(
+                new ConcurrentCircularQueue<Kernel32.INPUT_RECORD>(256));
+
+            // Task that will consume non-key events asynchronously
+            var consumeEvents = Task.Run(() =>
+            {
+                Console.WriteLine("consumeEvents started");
+
+                try
+                {
+                    while (!events.IsCompleted)
+                    {
+                        // blocking call
+                        var record = events.Take(token);
+
+                        Console.WriteLine("record: {0}",
+                            Enum.GetName(typeof(Kernel32.EVENT_TYPE), record.EventType));
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    // timeout
+                }
+
+                Console.WriteLine("consumeEvents ended");
+            }, token);
+
+            // Task that will watch for key events while feeding non-key events into our provider/consumer collection 
+            var readInputAndProduceEvents = Task.Run(async () =>
+            {
+                //So, this is the key point - we cannot use the following or we lose all non-key events:
+                // Stream stdin = Console.OpenStandardInput();
+                
+                // get a unicode character stream over console input 
+                Stream stdin = new ReadConsoleInputStream(handle, events);
+
+                // wrap in a System.IO.Pipelines.PipeReader to get clean async and span/memory usage 
+                var reader = StreamConnection.GetReader(stdin);
+
+                while (!token.IsCancellationRequested)
+                {
+                    // blocking call
+                    var result = await reader.ReadAsync(token);
+
+                    if (result.IsCanceled)
+                        break;
+
+                    var sequence = result.Buffer;
+                    var segment = sequence.Start;
+
+                    while (sequence.TryGet(ref segment, out var mem))
+                    {
+                        // decode back from unicode (2 bytes per char)
+                        var datum = Encoding.Unicode.GetString(mem.Span);
```

Comment:
do we know what happens here for a UTF-16 surrogate pair?

Reply:
It should just work -- I think the "2 bytes per char" comment is misleading: Mea culpa. The span that's being passed should be a complete representation of a utf-16 char, surrogate pairs and all.

Reply:
char, in this case, meaning a glyph.

Reply:
That said, isn't the NT console still kinda stuck with ucs-2? the key_event structure marshalls uChar as a WCHAR which is a fixed 16-bit value, right?

Reply:
I'm honestly not sure what happens with something like U+1F603 right now on the input queue side. It might show up as U+FFFD for a replacement, it might send key event type `INPUT_RECORD`s in a row with 0xD83D down, 0xDE03 down, 0xD83D up, 0xDE03 up... it might do 0xD83D down then 0xD83D up, 0xDE03 down, 0xDE03 up... it might also synthesize the sequence as an alt+numpad thing... I went to look at the source code briefly, and it's not any clearer to me. 

We have a work item somewhere in our tracker about actually defining this behavior and publishing how it works...

This is probably just undefined behavior for now and we can ignore my comment and just proceed assuming things are UCS-2 enough for now until proven otherwise.

Reply:
Hah, this was my thought process too. Except without the ability to go actually look at the source code :)